### PR TITLE
PID on sparks

### PR DESCRIPTION
**Pull request by Sid220**

### CODE REVIEW

**Comment by Sid220** [comment]
Tested in sim

File: src/main/java/frc/robot/subsystems/elevatorIO/ElevatorIOSparks.java [hunk 1]
```diff
@@ -87,11 +93,36 @@ public void updateInputs(ElevatorInputs inputs) {
...
+
+    currentLeftFF = leftFF.calculate(inputs.velocityInchesPerSecondLeft);
+    currentRightFF = rightFF.calculate(inputs.velocityInchesPerSecondRight);
   }
 
-  @Override
-  public void setVoltage(double volts) {
-    left.setVoltage(volts);
-    right.setVoltage(volts);
+  public void setPIDFF() {
```
Comment (tervay) [inline src/main/java/frc/robot/subsystems/elevatorIO/ElevatorIOSparks.java:101]: Does this actually get called anywhere?

File: src/main/java/frc/robot/subsystems/elevatorIO/ElevatorIOSparks.java [hunk 2]
```diff
@@ -87,11 +93,36 @@ public void updateInputs(ElevatorInputs inputs) {
     inputs.velocityInchesPerSecondRight = right.getEncoder().getVelocity();
     inputs.tempCelsiusRight = right.getMotorTemperature();
     inputs.currentDrawAmpsRight = right.getOutputCurrent();
+
+    currentLeftFF = leftFF.calculate(inputs.velocityInchesPerSecondLeft);
```
Comment (tervay) [inline src/main/java/frc/robot/subsystems/elevatorIO/ElevatorIOSparks.java:97]: Previously, we used the feedforward as part of our _velocity_ control loop - on the rio, we were generating a list of velocity setpoints for the motors to "chase," which if followed correctly, would land us in the right position. As a result, the feedforward needed to be based on our setpoint's velocity (not the current velocity like you have here).

We are not doing that on the Sparks. The equivalent on the Sparks would be `CANSparkMax.ControlType.kSmartMotion`, but it has some issues and we should stick with `CANSparkMax.ControlType.kPosition` (which is what you have). The difference is that the feedforward now needs to be based on the setpoint position, not the setpoint velocity.

You should set `currentLeftFF = leftFF.calculate(targetHeight);` in `goToSetpoint()`.

File: src/main/java/frc/robot/subsystems/swerveIO/module/SwerveModuleIOSparkMAX.java [hunk 3]
```diff
@@ -125,6 +154,17 @@ public void updateInputs(SwerveModuleInputs inputs) {
     inputs.driveOutputVolts = driver.getAppliedOutput() * RobotController.getBatteryVoltage();
     inputs.driveCurrentDrawAmps = driver.getOutputCurrent();
     inputs.driveTempCelcius = driver.getMotorTemperature();
+
+    azController.setReference(
+        aziSetpointDeg,
+        ControlType.kPosition,
+        0,
+        aziFF.calculate(inputs.aziEncoderVelocityDegPerSecond));
```
Comment (tervay) [inline src/main/java/frc/robot/subsystems/swerveIO/module/SwerveModuleIOSparkMAX.java:162]: feedforwards should be calculated based on setpoints, not current state

**Review by tervay — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/elevatorIO/ElevatorIOSparks.java (hunk 1 above)
Comment (Sid220) [reply src/main/java/frc/robot/subsystems/elevatorIO/ElevatorIOSparks.java:101]: Fixed in https://github.com/FRC2713/Robot2023/pull/77/commits/902c5a2d295ef635b6cba7075d53b0b53b197969

File: src/main/java/frc/robot/subsystems/elevatorIO/ElevatorIOSparks.java (hunk 2 above)
Comment (Sid220) [reply src/main/java/frc/robot/subsystems/elevatorIO/ElevatorIOSparks.java]: Fixed in 5fe8021b74669a15333238740e561b012b0ebede

**Pull request closed without merging**